core/tee/tee_sql_fs.c

```diff
@@ -0,0 +1,879 @@
+{
+	return tee_fs_rpc_rmdir(OPTEE_MSG_RPC_CMD_SQL_FS, name);
+}
+
+static int sql_fs_unlink_rpc(const char *file)
+{
+	return tee_fs_rpc_unlink(OPTEE_MSG_RPC_CMD_SQL_FS, file);
+}
+
+/*
+ * End of interface with tee-supplicant
+ */
+
+static size_t meta_size(void)
+{
+	return tee_fs_get_header_size(META_FILE) +
+	       sizeof(struct sql_fs_file_meta);
+}
+
+static size_t block_header_size(void)
+{
+	return tee_fs_get_header_size(BLOCK_FILE);
+}
+
+static size_t block_size_raw(void)
+{
+	return block_header_size() + BLOCK_SIZE;
+}
+
+/* Return the block number from a position in the user data */
+static ssize_t block_num(tee_fs_off_t pos)
+{
+	return pos / BLOCK_SIZE;
+}
+
+/* Return the block number from a position in the DB file */
+static ssize_t block_num_raw(tee_fs_off_t raw_pos)
+{
+	return (raw_pos - meta_size()) / block_size_raw();
+}
+
+/* Return the position of a block in the DB file */
+static ssize_t block_pos_raw(size_t block_num)
+{
+	return meta_size() + block_num * block_size_raw();
+}
+
+/* Given a position in the user data, return the offset in the DB file */
+static tee_fs_off_t pos_to_raw(tee_fs_off_t pos)
+{
+	tee_fs_off_t res;
+
+	if (pos < 0)
+		return -1;
+	res = meta_size() + block_num(pos) * block_size_raw();
+	if (pos % BLOCK_SIZE) {
+		res += block_header_size();
+		res += pos % BLOCK_SIZE;
+	}
+
+	return res;
+}
+
+/* Given a position in the DB file, return the offset in the user data */
+static tee_fs_off_t raw_to_pos(tee_fs_off_t raw_pos)
+{
+	tee_fs_off_t pos = raw_pos;
+	ssize_t n = block_num_raw(raw_pos);
+
+	if (n < 0)
+		return -1;
+
+	pos -= meta_size();
+	pos -= block_header_size();
+	if (pos < 0)
+		return -1;
+
+	return (n * BLOCK_SIZE) + (pos % BLOCK_SIZE);
+}
+
+static void put_fdp(struct sql_fs_fd *fdp)
+{
+	handle_put(&fs_db, fdp->fd);
+	free(fdp);
+}
+
+#ifndef CFG_ENC_FS
+static void copy_data(enum tee_fs_file_type type, uint8_t *raw_block,
+		      const uint8_t *data, size_t len)
+{
+	size_t header_size = tee_fs_get_header_size(type);
+
+	assert(type == META_FILE || type == BLOCK_FILE);
+
+	memset(raw_block, 0xFF, header_size);
+	memcpy(raw_block + header_size, data, len);
+}
+#endif
+
+static int write_meta(TEE_Result *errno, struct sql_fs_fd *fdp)
+{
+	int fd = fdp->fd;
+	size_t ct_size = meta_size();
+	uint8_t *ct;
+	int rc = -1;
+
+	ct = malloc(ct_size);
+	if (!ct) {
+		*errno = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit;
+	}
+
+	rc = sql_fs_lseek_rpc(fd, 0, TEE_FS_SEEK_SET);
+	if (rc < 0)
+		goto exit;
+
+#ifdef CFG_ENC_FS
+	{
+		TEE_Result res;
+
+		res = tee_fs_encrypt_file(META_FILE,
+					  (const uint8_t *)&fdp->meta,
+					  sizeof(fdp->meta), ct, &ct_size,
+					  fdp->meta.encrypted_fek);
+		if (res != TEE_SUCCESS) {
+			*errno = res;
+			goto exit;
+		}
+	}
+#else
+	copy_data(META_FILE, ct, (const uint8_t *)&fdp->meta,
+		  sizeof(fdp->meta));
+#endif
+
+	rc = sql_fs_write_rpc(fdp->fd, ct, ct_size);
+	if (rc != (int)ct_size)
+		rc = -1;
+	else
+		rc = 0;
+
+exit:
+	free(ct);
+	return rc;
```

Comment:
No, we'd better make sure `errno` is set to something different from `TEE_SUCCESS` each time `rc` is `< 0`. I think it's OK to initialize it to `TEE_ERROR_GENERIC` on entry, as it is done in `read_meta()`. Because the caller is supposed to check `errno` only in case `rc < 0`.